### PR TITLE
Introduce caching for X509 cerfificates

### PR DESCRIPTION
**Pull request by 0xxon**
This PR addresses #744 and introduces caching for X509 certificates. After a certificate is encountered a configurable amount of time, it is cached. When a certificate is cached, the core will no longer parse it - instead it will call a callback function in scriptland.

At the moment, the default callback functions just regenerates all the events as they would have been generated had the certificate been parsed again. This should have lower overhead than before - parsing does not actually happen and the Vals are recycled.

In the future, this could also be used to further customize behavior and to potentially not raise duplicate X509 events at all (which, however, certainly would break quite a few things).

### CODE REVIEW

File: src/file_analysis/analyzer/x509/X509.cc [hunk 1]
```diff
@@ -19,9 +19,7 @@
 #include <openssl/opensslconf.h>
 #include <openssl/err.h>
 
-namespace file_analysis {
-std::map<Val*, X509_STORE*> X509::x509_stores;
-}
+#include <iostream>
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/X509.cc]: That include may be leftover from debugging?

File: src/file_analysis/analyzer/x509/X509.cc [hunk 2]
```diff
@@ -45,10 +43,35 @@ bool file_analysis::X509::Undelivered(uint64_t offset, uint64_t len)
...
+			{
+			if ( ! cache_hit_callback )
+				return false;
+			// yup, let's call the callback.
+
+			val_list vl(2);
+			vl.push_back(GetFile()->GetVal()->Ref());
+			vl.push_back(new StringVal(cert_sha256));
+			Val* v = cache_hit_callback->Call(&vl);
+			Unref(v);
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/X509.cc]:
```suggestion
			IntrusivePtr<Val> v{AdoptRef{}, cache_hit_callback->Call(&vl)};
```

File: src/file_analysis/analyzer/x509/X509.h [hunk 3]
```diff
@@ -112,6 +113,18 @@ class X509 : public file_analysis::X509Common {
 	 */
 	static void FreeRootStore();
 
+	/**
+	 * Sets the table[string] that used as the certificate cache inside of Zeek.
+	 */
+	static void SetCertificateCache(IntrusivePtr<TableVal> cache)
+		{ certificate_cache = cache; }
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/X509.h]:
```suggestion
		{ certificate_cache = std::move(cache); }
```

File: src/file_analysis/analyzer/x509/X509.h [hunk 4]
```diff
@@ -112,6 +113,18 @@ class X509 : public file_analysis::X509Common {
...
+
+	/**
+	 * Sets the callback when a certificate cache hit is encountered
+	 */
+	static void SetCertificateCacheHitCallback(IntrusivePtr<Func> func)
+		{ cache_hit_callback = func; }
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/X509.h]:
```suggestion
		{ cache_hit_callback = std::move(func); }
```

File: src/file_analysis/analyzer/x509/functions.bif [hunk 5]
```diff
@@ -881,3 +881,44 @@ function x509_spki_hash%(cert: opaque of x509, hash_alg: count%): string
...
+##
+## .. note:: The base scripts use this function to set up certificate caching. You should only change the
+##           cache table if you are sure you will not conflict with the base scripts.
+##
+## .. zeek:see:: x509_set_certificate_cache_hit_callback
+function x509_set_certificate_cache%(tbl: string_any_table%) : bool
+	%{
+	file_analysis::X509::SetCertificateCache({NewRef{}, tbl->AsTableVal()});
+
+	return val_mgr->GetBool(1);
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/functions.bif]:
```suggestion
	return val_mgr->GetTrue();
```

File: src/file_analysis/analyzer/x509/functions.bif [hunk 6]
```diff
@@ -881,3 +881,44 @@ function x509_spki_hash%(cert: opaque of x509, hash_alg: count%): string
...
+##
+## .. note:: The base scripts use this function to set up certificate caching. You should only change the
+##           callback function if you are sure you will not conflict with the base scripts.
+##
+## .. zeek:see:: x509_set_certificate_cache
+function x509_set_certificate_cache_hit_callback%(f: string_file_hook%) : bool
+	%{
+	file_analysis::X509::SetCertificateCacheHitCallback({NewRef{}, f->AsFunc()});
+
+	return val_mgr->GetBool(1);
```
Comment (jsiwek) [inline src/file_analysis/analyzer/x509/functions.bif]:
```suggestion
	return val_mgr->GetTrue();
```

**Comment by 0xxon** [comment]
Oh, forgot the external test suite. Let me update that :)

File: scripts/base/files/x509/main.zeek [hunk 7]
```diff
@@ -50,13 +110,47 @@ event zeek_init() &priority=5
...
+	{
+	# we encountered a cached cert. The X509 analyzer will skip it. Let's raise all the events that it typically
+	# raises by ourselfes.
+
+	# first - let's checked if it already has an x509 record. That would mean that someone raised the file_hash event
+	# several times for the certificate - in which case we bail out.
+	if ( f$info?$x509 )
+		return;
+
+	local e = certificate_cache[sha256];
```
Comment (jsiwek) [inline scripts/base/files/x509/main.zeek]: Also just thought: this lookup already happened internally, so could potentially pass the result through via `hook` argument and avoid the 2nd lookup.  May use type `any` and cast if you want to keep API generic.

Comment (0xxon) [reply scripts/base/files/x509/main.zeek]: >     * Any idea of what the performance benefit vs. memory tradeoff is in either some real or benchmark measurement to confirm these changes have intended results?

Honestly, only a bit. The expected simultaneous number of certificates on larger links is a few 1,000. If we go with 5,000 and an average cert size of 50kb, this uses ~250mb of ram per worker process. Which is a not insignificant value.

For this you reduce the OpenSSL operations to basically 0 - and the number of allocations for certificates also goes way down. That being said - only testing this in a real-world environment will show if it is really worth it - I, however, expect that it is :)

This also opens the door to more drastic optimizations for sites that have really high TLS/certificate traffic - which can now just suppress events for duplicate certs.

Thanks for all the other suggestions - I will go and clean this up.

**Comment by 0xxon** [comment]
>     * Any idea of what the performance benefit vs. memory tradeoff is in either some real or benchmark measurement to confirm these changes have intended results?

Honestly, only a bit. The expected simultaneous number of certificates on larger links is a few 1,000. If we go with 5,000 and an average cert size of 50kb, this uses ~250mb of ram per worker process. Which is a not insignificant value.

For this you reduce the OpenSSL operations to basically 0 - and the number of allocations for certificates also goes way down. That being said - only testing this in a real-world environment will show if it is really worth it - I, however, expect that it is :)

This also opens the door to more drastic optimizations for sites that have really high TLS/certificate traffic - which can now just suppress events for duplicate certs.

Thanks for all the other suggestions - I will go and clean this up.

File: scripts/base/files/x509/main.zeek [hunk 8]
```diff
@@ -87,4 +196,32 @@ event file_state_remove(f: fa_file) &priority=5
...
+	if ( ! f?$info || "X509" !in f$info$analyzers || kind != "sha256" )
+		return;
+
+	if ( caching_required_encounters == 0 )
+		return;
+
+	if ( hash !in certificates_encountered )
+		certificates_encountered[hash] = 0;
+
+	certificates_encountered[hash] += 1;
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/base/files/x509/main.zeek]: Nope, it is not supposed to still keep track if it is already in the certificate_cache. Let me fix that.

File: src/file_analysis/analyzer/x509/X509.cc (hunk 1 above)
Comment (0xxon) [reply src/file_analysis/analyzer/x509/X509.cc]: indeed :)

**Comment by 0xxon** [comment]
And naturally forgot a test baseline update after the last change :)

**Comment by 0xxon** [comment]
Great, I will just merge it then.

**Pull request merged**